Dev healthcareapis microsoft.healthcare apis 2021 11 01 changes

dustinburson opened the pull request:
Adding new API version for 2021-11-01 stable.  This PR is taking the resources added in the 2021-06-01-preview and moving them to a stable version with some minor updates.

Changes include:
* Add User Assigned Managed Identity support for workspace related resources.
* Add Private Link resources for workspaces
* Remove referenced swaggers and add definitions into our Swagger. 
* Refactor readonly private link properties into a common definition for reuse
* Add Private Link properties for workspaces, fhirservices, and dicomservices
* Add EventState support to the FHIR Service
* Update Cors Origin Regex to address runaway condition.

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific langauge SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that label “WaitForARMFeedback” will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits.

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

dustinburson committed:
Adds base for updating Microsoft.HealthcareApis from version preview/2021-06-01-preview to version 2021-11-01

dustinburson committed:
Updates readme

dustinburson committed:
Updates API version in new specs and examples

dustinburson committed:
Initial updates for 2021-11-01 version

dustinburson committed:
Fix prettier errors and resolve systemData error

dustinburson committed:
Add missing type definitions

raosuhas commented:
>               "$ref": "#/definitions/PrivateEndpointConnection"

Why is this change made ? It is recommended to use the common types for the private links to ensure consistency across the azure resources. 

---
Refers to: specification/healthcareapis/resource-manager/Microsoft.HealthcareApis/stable/2021-11-01/healthcare-apis.json:471 in cbc165f. [](commit_id = cbc165f1f25748107a84f4dc3cb213f88cf16670, deletion_comment = False)

dustinburson commented:
> > ```
> >           "$ref": "#/definitions/PrivateEndpointConnection"
> > ```
> 
> Why is this change made ? It is recommended to use the common types for the private links to ensure consistency across the azure resources.
> 
> Refers to: specification/healthcareapis/resource-manager/Microsoft.HealthcareApis/stable/2021-11-01/healthcare-apis.json:471 in [cbc165f](https://github.com/Azure/azure-rest-api-specs/commit/cbc165f1f25748107a84f4dc3cb213f88cf16670). [](commit_id = [cbc165f](https://github.com/Azure/azure-rest-api-specs/commit/cbc165f1f25748107a84f4dc3cb213f88cf16670), deletion_comment = False)

The change was made to accommodate our RP validation framework.  We have a copy of our swagger deployed with our service that we use to validate incoming properties on our PUT and PATCH requests.  This is how we support [OAPI018](https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi018-reject-unknown-properties-and-property-values).  If we have the referenced types those types can't be validated.  We could maintain two versions of the Swagger one with the types self contained and one here with the references but that is undesirable due to the maintenance overhead and additional risk of mistakes when updating and synchronizing the content.

It also makes it easier to validate the Swagger with tooling since everything is self contained.  The data contracts are identical to the common references you pointed out.

raosuhas commented:
Cant you copy over the folder structure with the links to your repo ?. 

---
In reply to: [926043149](https://github.com/Azure/azure-rest-api-specs/pull/16108#issuecomment-926043149) [](http://example.com/codeflow?ancestors=926043149,926043149)

dustinburson commented:
> Cant you copy over the folder structure with the links to your repo ?.
> 
> In reply to: [926043149](https://github.com/Azure/azure-rest-api-specs/pull/16108#issuecomment-926043149)

Unfortunately that isn't the issue.  There are two problems.
* It isn't the presence in our repo that is the problem.  The swagger is used for runtime validation.  So all referenced files and their relative paths would need to be copied consistently for use in the service.
* The second, and larger issue is the Swagger validation framework we are using doesn't support referenced files.  I can revert the change but it means that we won't be supporting OAPI018 for the Private Link resources.

raosuhas commented:
The reason we ask to use common is also to avoid copy paste errors and to make it easier for the audience\reviewers to read this. Please note that this would come up everytime this comes up for ARM review. I would strongly suggest to update the swagger validation mechanism you use with one that supports these references. Again, that said , as long as you make sure that the resources have been copied exactly from the common , I am ok with signing off from ARM side for this change.

---
In reply to: [926074966](https://github.com/Azure/azure-rest-api-specs/pull/16108#issuecomment-926074966) [](http://example.com/codeflow?ancestors=926074966,926074966)

jianyexi commented on specification/healthcareapis/resource-manager/Microsoft.HealthcareApis/stable/2021-11-01/healthcare-apis.json:
```diff
@@ -0,0 +1,4087 @@
+            "$ref": "#/parameters/apiVersion"
+          },
+          {
+            "$ref": "#/parameters/subscriptionId"
```
could you keep the parameters as the order in the path ?

jianyexi replied:
This should apply to all the apis in the swagger

dustinburson replied:
@jianyexi  Thank you for the comment. Two follow up questions.
1. The apiVersion isn't a query parameter not a path parameter.  Does the comment still apply?
2. We have already published a Python SDK for Public Preview for these resources.  If I change the order of the API version I am assume it will reorder them in the SDKs generated causing breaking changes.  Is that not a concern?

jianyexi replied:
Okay, if the SDK already shipped , you can ignore my suggestion ,it only apply to new RP

jianyexi commented:
There are breaking changes in your PR against last stable version , you need to get approval through https://github.com/Azure/azure-rest-api-specs/pull/16108#issuecomment-984244823

https://github.com/Azure/azure-rest-api-specs/pull/16108#issuecomment-984244823

dustinburson commented:
> There are breaking changes in your PR against last stable version , you need to get approval through [#16108 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/16108#issuecomment-984244823)
> 
> [#16108 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/16108#issuecomment-984244823)

Thanks @jianyexi.  To be clear is the issue with the Cross-Version breaking changes, the Swagger SDK breaking changes or both?

dustinburson committed:
Remove pattern for validating Cors Origin.  Regex was found to have DDoS issues.  New correct pattern triggeres backwards breaking change alerts.  Removing patterns in latest iteration to avoid error.  Regexes are validated service side with updated logic already.

dustinburson committed:
Merge branch 'master' into dev-healthcareapis-Microsoft.HealthcareApis-2021-11-01-changes

dustinburson committed:
Add missing endtime from OperationResult

jianyexi commented:
> > There are breaking changes in your PR against last stable version , you need to get approval through [#16108 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/16108#issuecomment-984244823)
> > [#16108 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/16108#issuecomment-984244823)
> 
> Thanks @jianyexi Jianye Xi FTE. To be clear is the issue with the Cross-Version breaking changes, the Swagger SDK breaking changes or both?

I mean the cross-version breaking change

dustinburson committed:
Merge branch 'main' into dev-healthcareapis-Microsoft.HealthcareApis-2021-11-01-changes

dustinburson committed:
Revert CorsOriginEntry pattern removal to avoid false positive cross version breaking change

dustinburson commented:
> > > There are breaking changes in your PR against last stable version , you need to get approval through [#16108 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/16108#issuecomment-984244823)
> > > [#16108 (comment)](https://github.com/Azure/azure-rest-api-specs/pull/16108#issuecomment-984244823)
> > 
> > 
> > Thanks @jianyexi Jianye Xi FTE. To be clear is the issue with the Cross-Version breaking changes, the Swagger SDK breaking changes or both?
> 
> I mean the cross-version breaking change

@jianyexi I reverted the pattern change that was triggering the Cross Version Breaking Change errors.

dustinburson committed:
Add definition for Properties property in operation definition that was missing.

jianyexi commented:
@msyyc  to confirm with the PythonSDK breaking change

dustinburson committed:
Merge branch 'main' into dev-healthcareapis-Microsoft.HealthcareApis-2021-11-01-changes

dustinburson committed:
Merge branch 'main' into dev-healthcareapis-Microsoft.HealthcareApis-2021-11-01-changes

dustinburson committed:
Merge branch 'main' into dev-healthcareapis-Microsoft.HealthcareApis-2021-11-01-changes

msyyc commented:
Hi @dustinburson , some definition in swagger seems be removed compared with last tag. Is it expected?
![image](https://user-images.githubusercontent.com/70930885/145505645-ae0ca812-e9ac-4329-bc7a-78bc287ed34c.png)

msyyc commented on specification/healthcareapis/resource-manager/readme.md:
```diff
@@ -28,19 +28,28 @@ These are the global settings for HealthcareApis service.
 description: Azure Healthcare APIs Client
 openapi-type: arm
-tag: package-preview-2021-06
+tag: package-2021-01
```
Shall it be `package-2021-11`?

dustinburson replied:
This controls the default version of the API correct? If so this is intentional.  The API verison 2021-11-01 hasn't been published in ARM yet so we can't make this the default version.

dustinburson committed:
Test updating default verison per recommendation

dustinburson committed:
Merge branch 'main' into dev-healthcareapis-Microsoft.HealthcareApis-2021-11-01-changes

msyyc commented:
![image](https://user-images.githubusercontent.com/70930885/145740265-2da62ee5-1c4a-4e60-8c55-d83d7ac8b46c.png)
Now that `2021-11` has not been deployed, it is better not to change the default tag. I approve for Python breaking change since it is fake.

dustinburson committed:
Merge branch 'main' into dev-healthcareapis-Microsoft.HealthcareApis-2021-11-01-changes

dustinburson committed:
Update services and workspaces to use some provisioning state to remove .NET SDK error

dustinburson committed:
Resolve System.InvalidOperationException: Swagger document contains two or more x-ms-enum extensions with the same name 'ManagedServiceIdentityType' and different values: SystemAssigned,None vs. None,SystemAssigned,UserAssigned,SystemAssigned,UserAssigned

dustinburson committed:
Revert "Resolve System.InvalidOperationException: Swagger document contains two or more x-ms-enum extensions with the same name 'ManagedServiceIdentityType' and different values: SystemAssigned,None vs. None,SystemAssigned,UserAssigned,SystemAssigned,UserAssigned"

This reverts commit d44373c296808b50764b648279c60590a104a4b8.

dustinburson committed:
Resolve System.InvalidOperationException: Swagger document contains two or more x-ms-enum extensions with the same name 'ManagedServiceIdentityType' and different values: SystemAssigned,None vs. None,SystemAssigned,UserAssigned,SystemAssigned,UserAssigned"

dustinburson committed:
Merge branch 'main' into dev-healthcareapis-Microsoft.HealthcareApis-2021-11-01-changes

dustinburson committed:
Change default back to 2021-01-11 version

dustinburson committed:
Merge branch 'main' into dev-healthcareapis-Microsoft.HealthcareApis-2021-11-01-changes

dustinburson committed:
Change default version back in 2021-11-01 in anticipation of new release

dustinburson committed:
Revert changes that removed common-type references for local references

dustinburson committed:
Add async headers to patch examples

dustinburson committed:
Merge branch 'main' into dev-healthcareapis-Microsoft.HealthcareApis-2021-11-01-changes

dustinburson committed:
Add managed identity settings to dicomservices

dustinburson committed:
Add resourceVersionPolicyCOnfiguration to workspaces/fhirservices

dustinburson committed:
Fix prettier errors

dustinburson committed:
Add missing description for resourceTypeOverrides

dustinburson commented:
@jianyexi I have resolved all the required checks and they are now passing.  The few optional checks that are red I have reviewed and are intentional.  While you have signed off on the PR it is missing the Approved-OkToMerge required to commit the PR.  Can you please add the label?

jianyexi commented:
The only concern is the ApiReadiness error, could you ensure your service is deployed ?
https://github.com/Azure/azure-rest-api-specs/pull/16108/checks?check_run_id=4795378758

dustinburson commented:
@jianyexi the readiness check appears to be making a call to ARM looking for the operations resources under the 2021-11-01 API version. This failing is intentional, this API version contains resources that are in preview and will be released to GA at a later date. We need this PR complete so we can continue working on our GA items.

jianyexi commented:
@dustinburson  could you fix the following linting issues  in https://github.com/Azure/azure-rest-api-specs/pull/16108/checks?check_run_id=4795378650 , then I will add the label 'Approved-OkToMerge'
<html>
<body>
<!--StartFragment-->

Rule | Message
-- | --
❌ R4041 - XmsIdentifierValidation | Missing identifier id in array item propertyLocation: Microsoft.HealthcareApis/stable/2021-11-01/healthcare-apis.json#L2459
❌ R4041 - XmsIdentifierValidation | Missing identifier id in array item propertyLocation: Microsoft.HealthcareApis/stable/2021-11-01/healthcare-apis.json#L2932
❌ R4041 - XmsIdentifierValidation | Missing identifier id in array item propertyLocation: Microsoft.HealthcareApis/stable/2021-11-01/healthcare-apis.json#L2940
❌ R4041 - XmsIdentifierValidation | Missing identifier id in array item propertyLocation: Microsoft.HealthcareApis/stable/2021-11-01/healthcare-apis.json#L3016
❌ R4041 - XmsIdentifierValidation | Missing identifier id in array item propertyLocation: Microsoft.HealthcareApis/stable/2021-11-01/healthcare-apis.json#L3759

<!--EndFragment-->
</body>
</html>

dustinburson commented:
> 

@jianyexi this appears to be a new check (only started showing up in the past week or so).  I can't make the change requested because the underlying contract is already implemented in the service.  My understanding is the linting error is complaining because we have an array without an identifier field.  Making such a change now would break our contract.  I tried looking for examples for rule R4041 in the documentation but it isn't referenced here https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/openapi-authoring-automated-guidelines.md#R4041

If there is a way correct the linting error without changing the service implementation I will 

The ociArtifacts property is already implemented in the service.

The remaining definitions are also already implemented and define metrics for use in Azure Monitor and mirror implementations in other services (see https://github.com/Azure/azure-rest-api-specs/blob/b2f2423182b461f9b556ff68e9db7e93085eda7f/specification/network/resource-manager/Microsoft.Network/stable/2021-05-01/operation.json#L134) and have a contract we need to conform to.

Adding them to the Swagger now is just correcting omissions from our earlier swaggers.

jianyexi commented:
@dustinburson  sorry the link is here https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/openapi-authoring-automated-guidelines.md#r4041-xmsidentifiervalidation
fixing this rule will not break your contract, as it only need to add the 'x-ms-identifier' to the array definition, it has no impact to API behavior, it only add some metadata for the array definition

dustinburson commented:
> @dustinburson sorry the link is here https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/openapi-authoring-automated-guidelines.md#r4041-xmsidentifiervalidation fixing this rule will not break your contract, as it only need to add the 'x-ms-identifier' to the array definition, it has no impact to API behavior, it only add some metadata for the array definition

@jianyexi thanks for the example and clarification.  I will look at making the necessary changes.

dustinburson committed:
Merge branch 'main' into dev-healthcareapis-Microsoft.HealthcareApis-2021-11-01-changes

dustinburson committed:
Add missing  "x-ms-identifiers" property for arrays.

dustinburson commented:
@jianyexi I have added the xmsidentifers property and addressed the linting error.  They are now gone.

The Swagger Go Track 2 SDK check now has some failures.  These have been showing up intermittently when I have been addressing other issues and appear to be unrelated to the changes.  Is this a blocking issue (i.e. will prevent us from completing the PR?) and if so do you know who I can work with to understand where the errors are coming from?

dustinburson commented:
@jianyexi it appears the addition of the xmsidenitiers while solving the linting errors is now triggering SDK azure-sdk-for-net failures.  Please see https://github.com/Azure/azure-rest-api-specs/pull/16108/checks?check_run_id=4885819186

dustinburson committed:
Merge branch 'main' into dev-healthcareapis-Microsoft.HealthcareApis-2021-11-01-changes

dustinburson committed:
Merge branch 'main' into dev-healthcareapis-Microsoft.HealthcareApis-2021-11-01-changes
